controller/backup_controller.go

```diff
@@ -385,146 +400,164 @@ func (bc *BackupController) getBackupVolumeName(backup *longhorn.Backup) (string
+}
 
-	go func() {
-		state := backup.Status.State
-		defer func() {
-			backup, err := bc.ds.GetBackup(backup.Name)
-			if err != nil {
-				log.WithError(err).Errorf("Error get backup")
-				return
-			}
-			existingBackup := backup.DeepCopy()
+// syncWithMonitor syncs the backup state/progress from the replica monitor
+func (bc *BackupController) syncWithMonitor(backup *longhorn.Backup, volume *longhorn.Volume, monitor *engineapi.BackupMonitor) error {
+	if backup == nil || volume == nil || monitor == nil {
+		return nil
+	}
 
-			backup.Status.State = state
-			if reflect.DeepEqual(existingBackup.Status, backup.Status) {
-				return
-			}
-			if _, err := bc.ds.UpdateBackupStatus(backup); err != nil && !apierrors.IsConflict(errors.Cause(err)) {
-				log.WithError(err).Errorf("Error updating backup status")
-				return
-			}
-		}()
+	existingBackupState := backup.Status.State
 
-		if _, err = engineClient.SnapshotBackup(backup.Name, backup.Spec.SnapshotName, url, biName, biChecksum, backup.Spec.Labels, credential); err != nil {
-			state = longhorn.BackupStateError
-			event(err, state, backup, volume)
-			return
-		}
+	backupStatus := monitor.GetBackupStatus()
+	backup.Status.Progress = backupStatus.Progress
+	backup.Status.URL = backupStatus.URL
+	backup.Status.Error = backupStatus.Error
+	backup.Status.SnapshotName = backupStatus.SnapshotName
+	backup.Status.ReplicaAddress = backupStatus.ReplicaAddress
+	backup.Status.State = backupStatus.State
 
-		// Monitor snapshot backup progress
-		for {
-			engines, err := bc.ds.ListVolumeEngines(volume.Name)
-			if err != nil {
-				state = longhorn.BackupStateUnknown
-				event(err, state, backup, volume)
-				return
-			}
+	if existingBackupState == backup.Status.State {
+		return nil
+	}
 
-			bks := &longhorn.BackupStatus{}
-			for _, e := range engines {
-				backupStatusList := e.Status.BackupStatus
-				for _, b := range backupStatusList {
-					if b.SnapshotName == backup.Spec.SnapshotName {
-						bks = b
-						break
-					}
-				}
-			}
-			if bks == nil {
-				state = longhorn.BackupStateUnknown
-				event(err, state, backup, volume)
-				return
-			}
-			if bks.Error != "" {
-				state = longhorn.BackupStateError
-				event(errors.New(bks.Error), state, backup, volume)
-				return
-			}
+	if backup.Status.Error != "" {
+		bc.eventRecorder.Eventf(volume, corev1.EventTypeWarning, string(backup.Status.State),
+			"Snapshot %s backup %s label %v: %s", backup.Spec.SnapshotName, backup.Name, backup.Spec.Labels, backup.Status.Error)
+		return nil
+	}
+	bc.eventRecorder.Eventf(volume, corev1.EventTypeNormal, string(backup.Status.State),
+		"Snapshot %s backup %s label %v", backup.Spec.SnapshotName, backup.Name, backup.Spec.Labels)
 
-			if bks.Progress != 100 {
-				time.Sleep(BackupStatusQueryInterval)
-				continue
-			}
+	if backup.Status.State == longhorn.BackupStateCompleted {
+		return bc.syncBackupVolume(volume.Name)
+	}
```

Comment:
Moving this and the following monitor disabling into a separate function would make the flow more easy-to-understand. You can name the function something like `finishMonitoring` or `wrapUpMonitoring`.

Reply:
Put here is intended to only when the state changes to Completed, call `syncBackupVolume`.

Reply:
Change the below to a new function
```go
		switch backup.Status.State {
		case longhorn.BackupStateNew, longhorn.BackupStateInProgress:
			return nil
		case longhorn.BackupStateCompleted:
			bc.disableBackupMonitor(backup.Name)
		case longhorn.BackupStateError, longhorn.BackupStateUnknown:
			backup.Status.LastSyncedAt = syncTime
			bc.disableBackupMonitor(backup.Name)
			return nil
		}
```

would require an addition returned parameter to indicate whether to continue the reconcile loop _or_ immediately return it.

Reply:
OK. If it is not as simple as I expected, we can retain the current implementation. It's already much better than the very 1st version.